[exporter/elasticsearch] handle ecs mode mapping

j-kap-t opened the pull request:
Description:

Initial pass in implementing the ecs mapping mode that is documented but not implemented. This only covers a small subset of attribute mappings from otel > ecs, but the core record fields should be properly mapped.

Link to tracking Issue:

Testing:
Added test for ecs mapping mode.

Documentation:
Updated the documentation to reflect the default behavior is the current behavior with no mapping applied.

Reopening #30454

j-kap-t committed:
elasticsearchexporter: handle ecs mode mapping

Co-authored-by: Andrey Kaipov <andrey.kaipov@target.com>

j-kap-t committed:
Add changelogs

j-kap-t committed:
Fix test and lint error

j-kap-t committed:
Merge branch 'open-telemetry:main' into feat/ecs-format-m

j-kap-t committed:
Fix test

j-kap-t committed:
Merge branch 'main' into feat/ecs-format-m

j-kap-t committed:
Merge branch 'main' into feat/ecs-format-m

gregkalapos commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -41,16 +41,65 @@ const (
+		}
+
+		if n := record.SeverityNumber(); n != plog.SeverityNumberUnspecified {
+			document.AddInt("log.syslog.severity.code", int64(record.SeverityNumber()))
+			document.AddString("log.syslog.severity.name", record.SeverityText())
```
I think this part needs more thinking. ECS also has [`log.level`](https://www.elastic.co/guide/en/ecs/master/ecs-log.html#field-log-level) - if we deal with syslog, then this part would correctly map under `log.syslog.severity.*`, but if it's not syslog, then it should be under `log.level`.

ycombinator commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -41,16 +41,65 @@ const (
+		}
+
+		if n := record.SeverityNumber(); n != plog.SeverityNumberUnspecified {
+			document.AddInt("log.syslog.severity.code", int64(record.SeverityNumber()))
```
Similar to the comment about `log.syslog.severity.name`, `record.SeverityNumber()` should probably map to [`event.severity`](https://www.elastic.co/guide/en/ecs/current/ecs-event.html#field-event-severity) in the general (not syslog) case?

ycombinator commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -41,16 +41,65 @@ const (
+			document.AddString("log.syslog.severity.name", record.SeverityText())
+		}
+
+		document.AddAttribute("message", record.Body())
```
`record.Body()` could be a scalar or complex value, as documented in https://github.com/open-telemetry/opentelemetry-proto/blob/9d139c87b52669a3e2825b835dd828b57a455a55/opentelemetry/proto/logs/v1/logs.proto#L167-L170.  

>   // A value containing the body of the log record. Can be for example a human-readable
  // string message (including multi-line) describing the event in a free form or it can
  // be a structured data composed of arrays and maps of other values. [Optional].
  opentelemetry.proto.common.v1.AnyValue body = 5;

So I think this conversion needs to handle those different possibilities.  Alternatively, for now, you could just check if `record.Body().Type()` is `pcommon.ValueTypeStr` and if so set `message` to `record.Body().Str()`.  We can handle the other possibilities in a follow up PR.

ycombinator commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -41,16 +41,65 @@ const (
+		document.AddSpanID("span.id", record.SpanID())
+
+		if f := record.Flags(); f != 0 {
+			document.AddInt("trace.flags", int64(record.Flags()))
```
It's possible I overlooked something in ECS, but I'm not seeing a `trace.flags` field there: https://www.elastic.co/guide/en/ecs/current/ecs-tracing.html.

ycombinator commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -41,16 +41,65 @@ const (
+
+	switch m.mode {
+	case MappingECS:
+		document.AddTimestamp("@timestamp", record.Timestamp()) // We use @timestamp in order to ensure that we can index if the default data stream logs template is used.
```
I think this logic needs to be expanded, as per https://github.com/open-telemetry/opentelemetry-proto/blob/9d139c87b52669a3e2825b835dd828b57a455a55/opentelemetry/proto/logs/v1/logs.proto#L150-L153:

```suggestion
	if record.Timestamp() != 0 {
		document.AddTimestamp("@timestamp", record.Timestamp())
	} else {
		document.AddTimestamp("@timestamp", record.ObservedTimestamp())
	}
```

j-kap-t committed:
Address PR concerns

j-kap-t commented:
@ycombinator @gregkalapos thanks for the feedback. I hope I addressed the concerns. I wasn't the original dev on this internally, I'm just tying up some loose ends from someone who left us last year.

ycombinator commented:
Thanks @j-kap-t! I reviewed the changes and they look good to me.  I'm not an approver, however, so I believe @JaredTan95 will need to review and approve this PR to get it merged.

Also, thanks for the additional context re: tying up loose ends.  Appreciate you doing that!  I'm working on a follow up PR (#31694) where we can expand upon the foundation you're laying here for `mapping.mode: ecs`.

JaredTan95 commented on .chloggen/feat_ecs-format.yaml:
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
```
Others LGTM, But I have a question, is this a breaking for users who have already configured  with `ecs mode`?

j-kap-t replied:
@JaredTan95 thats actually a very good point. If a user explicitly has `ecs` specified in their config this would breaking change as the format of their logs would change. I will change the release notes to reflect this.

j-kap-t replied:
Fixed.

j-kap-t committed:
Clarify change is breaking if mapping is explicitly set to ecs

j-kap-t committed:
Merge branch 'main' into feat/ecs-format-m

j-kap-t committed:
Clarify changelog note

j-kap-t committed:
Merge branch 'main' into feat/ecs-format-m

j-kap-t committed:
Merge branch 'main' into feat/ecs-format-m

j-kap-t commented:
@codeboten Is there anything left needed for approval & merge? Thanks!

JaredTan95 commented:
@j-kap-t conflicts need to be solved.

j-kap-t committed:
Merge branch 'main' into feat/ecs-format-m

j-kap-t committed:
Fix formatting

j-kap-t commented:
@JaredTan95 fixed.

ycombinator commented:
@open-telemetry/collector-contrib-maintainer This PR is ready to merge, whenever you get a chance.

astencel-sumo commented on .chloggen/feat_ecs-format.yaml:
```diff
@@ -0,0 +1,27 @@
+note: Initial pass in implementing the `ecs` mapping mode
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [29742]
```
```suggestion
issues: [31553]
```

astencel-sumo commented on .chloggen/feat_ecs-format.yaml:
```diff
@@ -0,0 +1,27 @@
+component: elasticsearchexporter
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Initial pass in implementing the `ecs` mapping mode
```
Why is this an "initial pass"? Why not just "Implement the `ecs` mapping mode"?

j-kap-t replied:
The phrasing was because this only implements a very small subset of the full ECS mapping, like this intends to expand on: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31694.

astencel-sumo commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -41,20 +41,73 @@ const (
+		document.AddInt("SeverityNumber", int64(record.SeverityNumber()))
+		document.AddAttribute("Body", record.Body())
+		m.encodeAttributes(&document, record.Attributes())
+		document.AddAttributes("Attributes", record.Attributes())
```
Why is this line here? Doesn't it do the same thing as the above line `m.encodeAttributes(...)`, just not taking into account the mapping mode being `raw` or `none`?

j-kap-t replied:
Good catch, that was me failing to properly handle merge conflicts 🤦 .

astencel-sumo commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -41,20 +41,73 @@ const (
+
+		document.AddString("log.level", record.SeverityText())
+
+		if record.Body().Type() == pcommon.ValueTypeStr {
```
Oh I can see a previous comment about this: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31553#discussion_r1518405954.

I'm not sure if handing other body types in another PR is good enough, given that the component's stability is "Beta".

Can we at least create a separate issue to fix this after this PR is merged (hopefully soon)?

astencel-sumo replied:
Looking more into this, it seems there might be more changes coming in future PRs to how the `ecs` mapping mode works. Is this a reasonable assumption?

If yes, I propose to add a note in the README warning users that the `ecs` mapping mode is currently not stable. This note could be removed after this mode stabilizes.

What do you think @ycombinator @j-kap-t ?

ycombinator replied:
@astencel-sumo Yes, there is at least one follow up PR in the works for `ecs` mapping mode already: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31694.  And there might be others too in the future.  So I like the idea of adding a note to the `README` warning users that `ecs` mapping mode is experimental or something to that effect as part of this PR.

astencel-sumo commented on exporter/elasticsearchexporter/logs_exporter_test.go:
```diff
@@ -167,6 +168,40 @@ func TestExporter_PushEvent(t *testing.T) {
+		})(server.URL)
+		exporter := newTestExporter(t, server.URL, func(cfg *Config) { *cfg = *testConfig })
+		mustSendLogsWithAttributes(t, exporter,
+			// resource attrs
```
Looks like it's the other way around: first comes the record attributes map, second the resource attributes map.
```suggestion
			// record attrs
```

astencel-sumo commented on exporter/elasticsearchexporter/logs_exporter_test.go:
```diff
@@ -167,6 +168,40 @@ func TestExporter_PushEvent(t *testing.T) {
+				"application":  "myapp",
+				"service.name": "myservice",
+			},
+			// record attrs
```
```suggestion
			// resource attrs
```

j-kap-t committed:
Update .chloggen/feat_ecs-format.yaml

Co-authored-by: Andrzej Stencel <astencel@sumologic.com>

j-kap-t committed:
Apply suggestions from code review

Co-authored-by: Andrzej Stencel <astencel@sumologic.com>

j-kap-t committed:
Merge branch 'main' into feat/ecs-format-m

j-kap-t committed:
Fix merge conflict issue

astencel-sumo commented:
Thanks for the updates @j-kap-t . Can you also update the README, adding a note to the `ecs` mapping mode like " :warning: This mode's behavior is unstable, it is currently undergoing changes." or something similar?

j-kap-t committed:
Add warning indicating ecs mode is unstable
